doc: Explain effects of -prune=<n> parameter in release notes

xor-freenet opened the pull request:
As discussed in the mailing list thread

> [bitcoin-dev] Bitcoin Core 0.12.0 release candidate 1 available

in the replies to this message:
http://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-January/012276.html

Please review thoroughly, I'm a newbie.

xor-freenet committed:
doc: Explain effects of -prune=<n> parameter in release notes

As discussed in the mailing list thread:
    [bitcoin-dev] Bitcoin Core 0.12.0 release candidate 1 available
in the replies to this message:
http://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-January/012276.html

Please review thoroughly, I'm a newbie.

jonasschnelli commented on doc/release-notes.md:
```diff
@@ -267,6 +267,9 @@ Wallet: Pruning
 
 With 0.12 it is possible to use wallet functionality in pruned mode.
+This can reduce the disk usage from currently around 60 GB to
+less than 1 GB.
```
Around 2GB (also count the utxo set).

xor-freenet replied:
Thanks. Fixed by 46d7eb61372d4a507bf80924d1e19dafd46b8da1.

xor-freenet committed:
doc: Fix minimal disk usage with pruning enabled

petertodd commented on doc/release-notes.md:
```diff
@@ -274,6 +277,14 @@ To enable block pruning set `prune=<N>` on the command line or in
+the blockchain. In future releases, a higher value may also help the
+network as a whole: The stored blocks could be served to other nodes.
+Currently, nodes with pruning enabled do not serve block data at all
+and thus could be considered as "leechers".
```
I think we should remove that last sentence for now, as we're not in a position where "leaching" of blockchain data distribution bandwidth matters.

xor-freenet replied:
Thanks. Fixed by be4b474287d6a3479678b6c0ef50587f49bfbc54

jonasschnelli commented on doc/release-notes.md:
```diff
@@ -267,6 +267,9 @@ Wallet: Pruning
 
 With 0.12 it is possible to use wallet functionality in pruned mode.
+This can reduce the disk usage from currently around 60 GB to
+less than 2 GB.
```
I would prefer using "around 2 GB" to avoid an outdated document if utxo set, debug log, etc. grows.

xor-freenet replied:
Thanks. Fixed by 58e3abfffc69eaa5ea24f40f8428be48808ca664

xor-freenet committed:
doc: In release notes, do not claim that pruning is leeching

Peter Todd says it does not matter currently.

xor-freenet committed:
doc: In release notes, increase estimate of disk usage with pruning

Jonas Schnelli recommended this to account for growth of utxo set, debug
log, etc.

petertodd commented:
utACK https://github.com/xor-freenet/bitcoin/commit/58e3abfffc69eaa5ea24f40f8428be48808ca664

paveljanik commented on doc/release-notes.md:
```diff
@@ -274,6 +277,12 @@ To enable block pruning set `prune=<N>` on the command line or in
+wallet is as secure with high values as it is with low ones. Higher
+values merely reduce the network traffic in case of reorganization of
+the blockchain. In future releases, a higher value may also help the
+network as a whole: The stored blocks could be served to other nodes.
```
help the network as a whole: stored blocks...

xor-freenet replied:
Thanks. Fixed by a683d20d5f88f0685cdacdcba928d6ec572390dd

sdaftuar commented on doc/release-notes.md:
```diff
@@ -274,6 +277,12 @@ To enable block pruning set `prune=<N>` on the command line or in
 
+A value of 0 disables pruning. The minimal value above 0 is 550. Your
+wallet is as secure with high values as it is with low ones. Higher
+values merely reduce the network traffic in case of reorganization of
```
I think this is a confusing sentence.  Choosing a larger value means that a node can recover from larger reorgs; however the minimum value of 550MiB is designed to allow for reorgs up to 288 blocks (~2 days).

If there happened to be a reorg that required more data on disk than the user had (very unlikely) then the user's node will crash, and the user would have to restart the node with -reindex.  In the case of a node that was pruned, running -reindex will cause the node to delete all the data on disk, which can't (currently) be used by bitcoind when doing a -reindex.  So in particular the user will end up redownloading the whole blockchain in such a scenario (the amount re-downloaded will not depend on the value passed to -prune -- unless somehow the user selects a value large enough that all the block files remained on disk and no pruning actually took place).

Rather than try to re-explain pruning here, maybe it would make sense to refer readers back to the 0.11.0 release notes?  I don't know how easy that is to link to I guess.

laanwj replied:
I agree this misunderstands things a bit. The minimum value is chosen to handle any realistic reorganization. Larger reorgs will be disastrous for many other reasons, so there's no need to even consider them here.

If you want to link to older release notes the best way, I think, would be to link to the historical release notes in master: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.11.0.md#block-file-pruning . 
Alternatively you can link to the current release notes at the time of the v0.11.0 tag: 
https://github.com/bitcoin/bitcoin/blob/v0.11.0/doc/release-notes.md#block-file-pruning
These are probably the same but may potentially be missing some post-facto editorial fixes.

xor-freenet replied:
Thanks for your explanations. Fixed by 54d390780feaef895fede8da3f436b932af29c1d

With regards to the multiple options of resolving this, here's what I decided:

> Rather than try to re-explain pruning here, maybe it would make sense to refer readers back to the 0.11.0 release notes? I don't know how easy that is to link to I guess.

I think it makes sense to explain it here for the usability purpose of having a self-contained document. However, linking the old release notes also is useful. So I just did both.

> I agree this misunderstands things a bit. The minimum value is chosen to handle any realistic reorganization. Larger reorgs will be disastrous for many other reasons, so there's no need to even consider them here.

If the node can crash, that is worth mentioning at least? For example, it could cause financial loss for miners if their nodes shut down.
To find a tradeoff between your suggestion of not mentioning it at all, and the "crash + reindex required" information, I tried to keep it as neutral as possible by saying the node will "shutdown" (spelling of that word fixed in 15c0263ff1e4f4773005291e1eb584954900e197), and leaving out what has to be done afterwards. That both conveys the information that it won't work anymore, and allows leaving out what happens afterwards by not scaring the user off with a strong word such as "crash".

> If you want to link to older release notes the best way, I think, would be to link to the historical release notes in master: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.11.0.md#block-file-pruning .
Alternatively you can link to the current release notes at the time of the v0.11.0 tag:
https://github.com/bitcoin/bitcoin/blob/v0.11.0/doc/release-notes.md#block-file-pruning
These are probably the same but may potentially be missing some post-facto editorial fixes.

With regards to technical documentation, I think non-permanent links are an annoyance: They make it difficult to reconstruct what a user did read, since the document he saw will vary depending on when he accessed the link.
So I chose to link the tag.
Luckily, the two files are binary equivalent currently anyway.

laanwj commented:
Concept ACK

xor-freenet committed:
doc: Minor sentence length / capitalization fixes

xor-freenet committed:
doc: Fix wrong claims about blockchain reorganization with pruning

xor-freenet committed:
doc: Minor spelling fix

sdaftuar commented on doc/release-notes.md:
```diff
@@ -274,6 +277,17 @@ To enable block pruning set `prune=<N>` on the command line or in
+wallet is as secure with high values as it is with low ones. Higher
+values merely ensure that your node will not shut down upon blockchain
+reorganizations of more than 2 days - which are unlikely to happen in
+practice unless there is a hard fork. In future releases, a higher value
```
I would strike "unless there is a hard fork" from this sentence; hard forks shouldn't have much bearing on the likelihood of a 288 block reorg.

xor-freenet replied:
Thanks! Fixed by 8c5f90306c12142586024023785f20cac7b22055

sdaftuar commented:
Looks better -- utACK apart from my one comment.

xor-freenet committed:
doc: In release notes, reduce length of pruning section

laanwj commented:
Thanks, looks good to me now, going to merge.
